src/NBitcoin/BlockValidator.cs

```diff
@@ -1,940 +0,0 @@
-            int actualSpacing = (int)(pindexPrev.Header.Time - pindexPrevPrev.Header.Time);
-            if (IsProtocolV1RetargetingFixed(indexLast.Height))
-            {
-                if (actualSpacing < 0) actualSpacing = targetSpacing;
-            }
-            if (IsProtocolV3((int)indexLast.Header.Time))
-            {
-                if (actualSpacing > targetSpacing * 10) actualSpacing = targetSpacing * 10;
-            }
-
-            // target change every block
-            // retarget with exponential moving toward target spacing
-            var targetTimespan = 16 * 60; // 16 mins
-            var target = pindexPrev.Header.Bits.ToBigInteger();
-
-            int interval = targetTimespan / targetSpacing;
-            target = target.Multiply(BigInteger.ValueOf(((interval - 1) * targetSpacing + actualSpacing + actualSpacing)));
-            target = target.Divide(BigInteger.ValueOf(((interval + 1) * targetSpacing)));
-
-            if (target.CompareTo(BigInteger.Zero) <= 0 || target.CompareTo(targetLimit) >= 1)
-                //if (target <= 0 || target > targetLimit)
-                target = targetLimit;
-
-            return new Target(target);
-        }
-
-        public static bool CheckBlockSignature(Block block)
-        {
-            if (BlockStake.IsProofOfWork(block))
-                return block.BlockSignatur.IsEmpty();
-
-            if (block.BlockSignatur.IsEmpty())
-                return false;
-
-            var txout = block.Transactions[1].Outputs[1];
-
-            if (PayToPubkeyTemplate.Instance.CheckScriptPubKey(txout.ScriptPubKey))
-            {
-                var pubKey = PayToPubkeyTemplate.Instance.ExtractScriptPubKeyParameters(txout.ScriptPubKey);
-                return pubKey.Verify(block.GetHash(), new ECDSASignature(block.BlockSignatur.Signature));
-            }
-
-            if (IsProtocolV3((int)block.Header.Time))
-            {
-                // Block signing key also can be encoded in the nonspendable output
-                // This allows to not pollute UTXO set with useless outputs e.g. in case of multisig staking
-
-                var ops = txout.ScriptPubKey.ToOps().ToList();
-                if (!ops.Any()) // script.GetOp(pc, opcode, vchPushValue))
-                    return false;
-                if (ops.ElementAt(0).Code != OpcodeType.OP_RETURN) // OP_RETURN)
-                    return false;
-                if (ops.Count < 2) // script.GetOp(pc, opcode, vchPushValue)
-                    return false;
-                var data = ops.ElementAt(1).PushData;
-                if (!ScriptEvaluationContext.IsCompressedOrUncompressedPubKey(data))
-                    return false;
-                return new PubKey(data).Verify(block.GetHash(), new ECDSASignature(block.BlockSignatur.Signature));
-            }
-
-            return false;
-        }
-
-        public static bool IsCanonicalBlockSignature(Block block, bool checkLowS)
```

Comment:
This is something that I found in stratisX and it is not explicitly used in C# node, so I was wondering if it is implicitly checked with the current signature verification. But seeing that here, maybe we are missing this check compared to stratisX

Reply:
Yeah we have a TODO rule in place for this so I did not forget.
https://github.com/stratisproject/StratisBitcoinFullNode/blob/master/src/Stratis.Bitcoin.Features.Consensus/Rules/CommonRules/PosBlockContextRule.cs

That's why I feel its safe to delete from NBitcoin